packages/flutter/lib/src/painting/shader_warm_up.dart

```diff
@@ -0,0 +1,144 @@
+/// --type=skia` and analyze it with https://debugger.skia.org. Alternatively,
+/// one may run the app with `flutter run --trace-skia` and then examine the GPU
+/// thread in the observatory timeline to see which Skia draw operations are
+/// commonly used, and which shader compilations are causing janks.
+///
+/// The size of [width]=[height]=1000 here are picked arbitrarily. The only goal
+/// is to have them not too far away from the actual resolution (very few mobile
+/// devices seem to have a resolution that is 10x larger or 10x smaller than
+/// 1000x1000). The size should have little effect on the shader warm-up. A
+/// custom shader warm up can pick another arbitrary number that's close to the
+/// targeted devices.
+class DefaultShaderWarmUp extends ShaderWarmUp {
+  /// Allow [DefaultShaderWarmUp] to be used as the default value of parameters.
+  const DefaultShaderWarmUp();
+
+  @override
+  int get width => 1000;
+
+  @override
+  int get height => 1000;
+
+  /// Trigger common draw operations on a canvas to warm up GPU shader
+  /// compilation cache.
+  @override
+  void warmUpOnCanvas(Canvas canvas) {
+    final RRect rrect = RRect.fromLTRBXY(20, 20, 60, 60, 10, 10);
+    final Path rrectPath = Path()..addRRect(rrect);
+
+    final Path circlePath = Path()..addOval(
+        Rect.fromCircle(center: const Offset(40, 40), radius: 20)
+    );
+
+    // The following path is copied from
+    // https://skia.org/user/api/SkCanvas_Reference#SkCanvas_drawPath
+    final Path path = Path();
+    path.moveTo(20, 60);
+    path.quadraticBezierTo(60, 20, 60, 60);
+    path.close();
+    path.moveTo(60, 20);
+    path.quadraticBezierTo(60, 60, 20, 60);
+
+    final List<Path> paths = <Path>[rrectPath, circlePath, path];
+
+    final List<Paint> paints = <Paint>[
+      Paint()
+        ..isAntiAlias = true
+        ..style = PaintingStyle.fill,
+      Paint()
+        ..isAntiAlias = true
+        ..style = PaintingStyle.stroke
+        ..strokeWidth = 10,
+      Paint()
+        ..isAntiAlias = true
+        ..style = PaintingStyle.stroke
+        ..strokeWidth = 0.1  // hairline
+    ];
+
+    // Warm up path stroke and fill shaders.
+    for (int i = 0; i < paths.length; i++) {
+      canvas.save();
+      for (Paint paint in paints) {
+        canvas.drawPath(paths[i], paint);
+        canvas.translate(80, 0);
+      }
+      canvas.restore();
+      canvas.translate(0, 80);
+    }
+
+    // Warm up shadow shaders.
+    const Color black = Color(0xFF000000);  // foundation can't access material/colors
+    canvas.save();
+    canvas.drawShadow(rrectPath, black, 10.0, true);
+    canvas.translate(80, 0);
+    canvas.drawShadow(rrectPath, black, 10.0, false);
+    canvas.restore();
+
+    // Warm up text shaders.
+    canvas.translate(0, 80);
+    final ParagraphBuilder paragraphBuilder = ParagraphBuilder(
+      ParagraphStyle(textDirection: TextDirection.ltr),
+    )..pushStyle(TextStyle(color: black))..addText('Hello, world.');
```

Comment:
let's pick some more minimal text like the empty string or a space or an emoji or something that isn't quite so arbitrary as "Hello world".